cpp/src/join/hash_join.cu

```diff
@@ -43,51 +61,38 @@ std::pair<std::unique_ptr<table>, std::unique_ptr<table>> get_empty_joined_table
- *
  * @param build Table of columns used to build join hash.
+ * @param hash_table Build hash table.
  * @param compare_nulls Controls whether null join-key values should match or not.
  * @param stream CUDA stream used for device memory operations and kernel launches.
  *
- * @return Built hash table.
  */
-std::unique_ptr<multimap_type, std::function<void(multimap_type*)>> build_join_hash_table(
-  cudf::table_view const& build, null_equality compare_nulls, rmm::cuda_stream_view stream)
+void build_join_hash_table(cudf::table_view const& build,
+                           multimap_type& hash_table,
+                           null_equality compare_nulls,
+                           rmm::cuda_stream_view stream)
 {
   auto build_device_table = cudf::table_device_view::create(build, stream);
 
   CUDF_EXPECTS(0 != build_device_table->num_columns(), "Selected build dataset is empty");
   CUDF_EXPECTS(0 != build_device_table->num_rows(), "Build side table has no rows");
 
-  size_type const build_table_num_rows{build_device_table->num_rows()};
-  std::size_t const hash_table_size = compute_hash_table_size(build_table_num_rows);
+  row_hash hash_build{*build_device_table};
+  auto const empty_key_sentinel = hash_table.get_empty_key_sentinel();
+  make_pair_function pair_func{hash_build, empty_key_sentinel};
 
-  auto hash_table = multimap_type::create(hash_table_size,
-                                          stream,
-                                          true,
-                                          multimap_type::hasher(),
-                                          multimap_type::key_equal(),
-                                          multimap_type::allocator_type());
+  thrust::counting_iterator<size_type> first(0);
+  auto iter = thrust::make_transform_iterator(first, pair_func);
 
-  row_hash hash_build{*build_device_table};
-  rmm::device_scalar<int> failure(0, stream);
-  constexpr int block_size{DEFAULT_JOIN_BLOCK_SIZE};
-  detail::grid_1d config(build_table_num_rows, block_size);
-  auto const row_bitmask = (compare_nulls == null_equality::EQUAL)
-                             ? rmm::device_buffer{0, stream}
-                             : cudf::detail::bitmask_and(build, stream);
-  build_hash_table<<<config.num_blocks, config.num_threads_per_block, 0, stream.value()>>>(
-    *hash_table,
-    hash_build,
-    build_table_num_rows,
-    static_cast<bitmask_type const*>(row_bitmask.data()),
-    failure.data());
-  // Check error code from the kernel
-  if (failure.value(stream) == 1) { CUDF_FAIL("Hash Table insert failure."); }
-
-  return hash_table;
+  size_type const build_table_num_rows{build_device_table->num_rows()};
+  if (compare_nulls == null_equality::EQUAL) {
+    hash_table.insert(iter, iter + build_table_num_rows, stream.value());
+  } else {
+    auto const row_bitmask = cudf::detail::bitmask_and(build, stream);
+    row_contains_null pred{static_cast<bitmask_type const*>(row_bitmask.data())};
```

Comment:
There's an opportunity for optimization here where we don't need to build the `row_bitmask` when none of the columns in the build table contain nulls. 
```suggestion
  if ((compare_nulls == null_equality::EQUAL) or (not nullable(build))) {
    hash_table.insert(iter, iter + build_table_num_rows, stream.value());
  } else {
    auto const row_bitmask = cudf::detail::bitmask_and(build, stream);
    row_contains_null pred{static_cast<bitmask_type const*>(row_bitmask.data())};
```